### ISSUE TITLE
Drag Sorting issues in Web Component with Shadow Dom disabled

### ISSUE THREAD

**Issue by jenter** [artf/grapesjs#2999]
_Version 16.10_

**Are you able to reproduce the bug from the demo?**

[X] Yes
[ ] No

**What is the expected behavior?**

That GrapesJS can support dragging & sorting interaction with custom Web Components which do not have the Shadow DOM enabled.

**Describe the bug**

Hello, we use GrapesJS extensively and integrate many of our custom web components, which are generated using the [StencilJS](https://stenciljs.com/) project. An issue that we run into is that dragging elements into the Web Components. However, this is only problematic if the Shadow DOM is disabled (`shadow: false`). 

Shadow DOM enabled: [No Bug]

![ShadowON](https://user-images.githubusercontent.com/1368216/92161326-780fd100-edfe-11ea-9826-c697750011f5.gif)

Shadow DOM disabled: [Bug]

![ShadowOFF](https://user-images.githubusercontent.com/1368216/92161332-7ba35800-edfe-11ea-9d40-55adc4bc3575.gif)

The issue is that when you are attempting to drop an HTML element in the Web Component, Grapes doesn't seem to understand the DOM / Node references (assuming) and therefore not able to drop within. What's odd is that this bug is not an issue (even with shadow dom disabled) if you wrap any slotted content in a DIV. 
```
# Bug 
<faux-grid-item>
      <p>Example</p>
       ...
```
```
# No Bug 
<faux-grid-item>
   <div>
      <p>Example</p>
       ...
```

This bug/request isn't just for StencilJS, but just in general with Web Components and the Shadow DOM. 
Are there best practices with Grapes and Web Components with enabling or tracking events to mediate these issues?

Example from screenshot of custom component in stenciljs:
```
@Component({
  tag: 'faux-grid-item',
  styleUrl: 'faux-grid-item.css',
  shadow: false, // <<=== causes issues
})
```

**Are you able to attach screenshots, screencasts or a live demo?**

[X] Yes (attach)
[ ] No

**Comment by artf** [comment]
> This bug/request isn't just for StencilJS, but just in general with Web Components and the Shadow DOM.
Are there best practices with Grapes and Web Components with enabling or tracking events to mediate these issues?

I'm not working with Web Components that much so if you're able to create a reproducible demo (eg. custom component using Web Components under the hood) that would help us to debug and understand the issue.

**Comment by jenter** [comment]
@artf sure thing, thanks for the quick response. 

I have imported my custom web component library here in this Fiddle to demonstrate the issue: https://jsfiddle.net/jenter77/5btn2ao7/25/

![fiddleSort](https://user-images.githubusercontent.com/1368216/92647514-ff08f180-f2b5-11ea-8433-697820d85fe3.gif)

If you reference my public repo and example "faux-grid-item" web component [here](https://github.com/jenter/jenter-stenciljs-starter/blob/master/src/components/faux-grid-item/faux-grid-item.tsx) you will see it's just a simple component that delays for a couple of seconds and then loads. If shadowDOM is off the sorting/dragging issue bug occurs. 
```
import { Component, State, h } from '@stencil/core';

@Component({
  tag: 'faux-grid-item',
  styleUrl: 'faux-grid-item.css',
  shadow: false, // <<=== causes issues
})

...
```

**Comment by artf** [comment]
Hi Jason, I've tried to give a look at what is happening inside the editor and, I'm not sure how the `shadow` option actually affects this behavior but, the main issue I see is that dynamic `<div class="delayed--wrapper">` element.
Basically the editor, post parsing, has this component structure
```html
<faux-grid-item>
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
</faux-grid-item>
```
but the actual DOM is
```html
<faux-grid-item>
   <div class="delayed--wrapper">
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
   </div>
</faux-grid-item>
```
This misalignment confuses the sorter because `div.delayed--wrapper` is not connected to any component. 
Indeed, to "fake" your expected behavior you would need to keep one div inside your template
```html
<faux-grid-item>
   <div>
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
   </div>
</faux-grid-item>
```

Unfortunately, I don't see how this can be solved (the reason why I'll close the issue). 
When the sorter tells to drop some Block at position X (eg. after the 2 children) of the FauxGridItem, it will append the new dropped component inside the FauxGridItem's children collection and then the component's view renders the DOM at the same position, but in this case, it won't be possible because FauxGridItem has only 1 child (`div.delayed--wrapper`). 

So to recap, modifying the DOM outside of the editor creates unpredictable side effects. 
It's the same with frameworks which rely on the virtual DOM (eg. React, Vue, etc.) it's always recommended to avoid dealing directly with the DOM as you might break its virtual state.

**Comment by jenter** [comment]
@artf thanks for the feedback. Are there any suggestions on how to fire events or attach listeners to/from Grapes that would help in tracking any DOM changes within web components that update later in the cycle?

**Comment by artf** [comment]
it won't work because you will probably end up having components you wouldn't expect. 

For example, you will trigger the parser post-DOM update, then in the export code instead of
```html
<faux-grid-item>
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
</faux-grid-item>
```
you will have
```html
<faux-grid-item>
   <div class="delayed--wrapper">
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
   </div>
</faux-grid-item>
```
so then (in the published page), your StencilJS code will generate this
```html
<faux-grid-item>
  <div class="delayed--wrapper">
    <div class="delayed--wrapper">
        <p>Hello</p>
        <p>Hello</p>
        <p>Hello</p>
    </div>
  </div>
</faux-grid-item>
```
I hope it's clear. The only reasonable thing I can think of use of custom components (eg. grapesjs MJML plugin was made in that way)

**Comment by jenter** [comment]
@artf thanks for the quick feedback & this is very helpful in our next steps forward. 
So am I correct in saying: any updates or events should be handled in the webcomponent itself instead of focusing those efforts within any type of GrapesJS binding / firing?

**Comment by artf** [comment]
Not really, you're not able to guarantee the full coherency if you have an external script that manipulates the DOM of components. My advice would be to create custom components in GrapesJS

**Comment by jenter** [comment]
@artf makes sense, thank you.
